staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -1219,6 +1222,237 @@ func RunSendInitialEventsBackwardCompatibility(ctx context.Context, t *testing.T
 	w.Stop()
 }
 
+func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interface) {
+	trueVal, falseVal := true, false
+	scenarios := []struct {
+		name                string
+		allowWatchBookmarks bool
+		sendInitialEvents   *bool
+		resourceVersion     string
+
+		initialPods                func(ns string) []*example.Pod
+		podsAfterEstablishingWatch func(ns string) []*example.Pod
+
+		expectedInitialEventsInStrictOrder   func(ns string) []watch.Event
+		expectedInitialEventsInRandomOrder   func(ns string) []watch.Event
+		expectedEventsAfterEstablishingWatch func(ns string) []watch.Event
+	}{
+		{
+			name:                "allowWatchBookmarks=true, sendInitialEvents=true, RV=0",
+			allowWatchBookmarks: true,
+			sendInitialEvents:   &trueVal,
+			resourceVersion:     "0",
+			initialPods:         func(ns string) []*example.Pod { return []*example.Pod{makePod(ns, "2"), makePod(ns, "3")} },
+			expectedInitialEventsInRandomOrder: func(ns string) []watch.Event {
+				return []watch.Event{
+					{Type: watch.Added, Object: makePodWithRV(ns, 2)},
```

Comment:
This is making an assumption about how storage is assigning RV, which is tricky.

We should figure out to make it RV-independent (especially given you're later generating pods with consecutive RVs (so effectively make the test cases dependent on the earlier ones).

I think that something like that should work:
(1) add a field `createdInitialPods []*example.Pod` to scenario
(2) this will be propagated by the test itself (when creating those)
(3) the function here will effectively be something like:
```
expectedInitialEventsInRandomOrder: func(ns string) []watch.Event {
  {Type: watch.Added, Object: createdPods[0]},
  ...
```

Reply:
We probably need something similar for podsAfterEstablishingWatch

Reply:
these test are low-level and are intended to make assumption about the underlying storage.
having an rv is easier to read and debug potential issues i don't see anything wrong with that.

Reply:
You have no guarantee about RVs for etcd - what I'm suggesting will automatically make it work also for etcd.
And it's not introducing visible complexity.

I'm pretty convinced that what I'm proposing is actually better.

Reply:
> You have no guarantee about RVs for etcd

I have otherwise the whole streaming doesn't make sense because we wouldn't be able to compare RVs

Reply:
That's not true, etcd guarantees monotonical increase, but not exact values.
If etcd would suddenly decide to incrase RV by 2 on every transaction, this would still be correct, my proposal would work and yours would start failing.

Reply:
The test would still be correct because it assumes that RVs are monotonically increasing.

Reply:
No - you assume the exact RVs assigned by etcd to be 2, 3, 4, 5, ... etc

Reply:
ok, updated, ptal.